pkg/registry/resource/devicetaintrule/storage/storage_test.go

```diff
@@ -53,8 +54,9 @@ func validNewDeviceTaint(name string) *resource.DeviceTaintRule {
 		},
 		Spec: resource.DeviceTaintRuleSpec{
 			Taint: resource.DeviceTaint{
-				Key:    "example.com/taint",
-				Effect: resource.DeviceTaintEffectNoExecute,
+				Key:       "example.com/taint",
+				Effect:    resource.DeviceTaintEffectNoExecute,
+				TimeAdded: &metav1.Time{Time: time.Now().Truncate(time.Second)}, // Must know in advance what will be stored.
```

Comment:
This seems fine, but I'm wondering why without this PR the test was flaking relatively rarely.
Why it wasn't consistently failing?

Reply:
Basically, the created objects here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go#L1392-L1393

had never the field set, whereas the ones that we got through list should always be defaulted, no?

So how they could be the same?

Reply:
> The object without TimeAdded is not being compared.

That's what I don't yet understand - for list tests it seems that exactly those are compared:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go#L1376
+
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/rest/resttest/resttest.go#L1392-L1393

Reply:
Actually - these objects are modified here (when creation happens):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/testing/utils.go#L46-L56

Reply:
> But it doesn't happen only once! Here's one backtrace of several for a single TestCreate:

Hmm - I don't follow what the stacktrace shows - default doens't happen twice there?

Reply:
I think I know what's happening here:
- we're creating object that doesn't have the field set (see my comment above how it's created)
- this is only defaulted on conversion back so on reading
- when listing data, defaulting is happening again (because the object in storage is not defaulted)

This seems like a test issue that we should fix, but we should do that as a follow-up.

Reply:
> I don't follow what the stacktrace shows - default doens't happen twice there?

I only gave one such backtrace. `SetDefaults_DeviceTaint` is hit multiple times with a similar backtrace.

> I think I know what's happening here: [...]

Yes, that is also what I suspected.

> This seems like a test issue that we should fix, but we should do that as a follow-up.

=> https://github.com/kubernetes/kubernetes/issues/130975